libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -1620,6 +1623,30 @@ bool Compass::configured(char *failure_msg, uint8_t failure_msg_len)
     return all_configured;
 }
 
+void Compass::_reset_compass_id()
+{
+#if COMPASS_MAX_INSTANCES > 1
+    // Check if any of the registered devs are not registered
+    for (Priority i(0); i<COMPASS_MAX_INSTANCES; i++) {
+        if (_get_state(i).dev_id == 0) {
+            _priority_did_stored_list[i].set_and_save(0);
+        }
+    }
+    
+    // Check if any of the registered devs are not registered
+    for (StateIndex i(0); i<COMPASS_MAX_INSTANCES; i++) {
+        if (_state[i].dev_id == 0) {
+            // also hard reset dev_ids that are not detected
+            _state[i].dev_id.save();
+            gcs().send_text(MAV_SEVERITY_ALERT, "Mag: Compass #%d with DEVID %lu deregistered!", uint8_t(i), (unsigned long)_state[i].expected_dev_id);
+        }
+    }
+
+    // Lets do init one more time.
+    init();
```

Comment:
this calls the init functions again in the backends, which causes spurious errors and bus failures (eg. makes AK09916 on CubeOrange fail)

Reply:
@tridge register_periodic_callback should not be called again, as calling register_compass method inside init calls second time should fail. I am trying and see if I can reproduce this.

Reply:
This issue was actually due to failure to reinitialise auxiliary bus over SPI. I have modified the driver to only try reinitialise external I2C and UAVCAN compasses. I have tested using Here Compass in I2C and UAVCAN mode. When we are going through I2C reinit the probe is not called for device with id already used, and similar mechanism is in place for UAVCAN as well. Also on top of that I have added code in register compass to return false if the call is being done for dev_id already registered.